### DISCUSSION TITLE
[Help]: Adding multiple plugins with Vue

### DISCUSSION THREAD

**Discussion by richgcook** [davidjerleke/embla-carousel#853] [Help] [answered]
### Summary

Is this a working way to set up multiple plugins when using emblaCarouselVue?

```
import emblaCarouselVue from 'embla-carousel-vue'
import AutoScroll from 'embla-carousel-auto-scroll'
import ClassNames from 'embla-carousel-class-names'

const [viewport, embla] = emblaCarouselVue({
	loop: true,
	align: 'start',
	dragFree: true,
	containScroll: 'trimSnaps',
}, [
	ClassNames({
		draggable: '--draggable', 
		dragging: '--dragging'
	}),
	AutoScroll(
		{ playOnInit: false }
	)]
)
```

It's not the cleanest but technically it should work? I had previously used the following but unsure which is cleaner.

```
const classNamesOptions = { draggable: '--draggable', dragging: '--dragging' }
const classNamesWithOptions = ClassNames(classNamesOptions)

const [viewport, embla] = emblaCarouselVue({
	loop: true,
	align: 'start',
	dragFree: true,
	containScroll: 'trimSnaps',
}, [classNamesWithOptions])
```

### If applicable, which variants of Embla Carousel are relevant to this question?

- [ ] embla-carousel **(Core)**
- [ ] embla-carousel-react
- [X] embla-carousel-vue
- [ ] embla-carousel-svelte
- [ ] embla-carousel-autoplay
- [ ] embla-carousel-auto-scroll
- [ ] embla-carousel-solid
- [ ] embla-carousel-auto-height
- [ ] embla-carousel-class-names
- [ ] embla-carousel-docs **(Documentation)**
- [ ] embla-carousel-docs **(Generator)**

**Comment by davidjerleke** [comment]
@richgcook as stated in [the docs](https://www.embla-carousel.com/api/plugins/#constructor-options) (please click on the Vue tab for relevant code example) the first approach you mention is correct. You should pass the plugin constructors directly to `emblaCarouselVue()`.

However, if you want to make it a bit more readable, you can store the plugin options in a variable like so:

```html
<script setup>
  import emblaCarouselVue from 'embla-carousel-vue'
  import Autoplay from 'embla-carousel-autoplay'

  const autoplayOptions = { delay: 4000 }

  const [emblaRef] = emblaCarouselVue({ loop: true }, [
    Autoplay(autoplayOptions)
  ])

  // ...
</script>
```

**Reply by richgcook** [reply]
Thanks so much for helping. I've tried this, and actually tried implementing my original approach (the second example in my first post) which worked on an older project but on a new project neither approach works. Is there a Vue sandbox available to test?

**Reply by davidjerleke** [reply]
@richgcook please create a CodeSandbox with your setup so I can debug what’s going on.

**Reply by richgcook** [reply]
Do you have one set up for Vue/Nuxt through any previous issues/tests? If not I'll create one later and add it in here. Thanks, again.

**Reply by davidjerleke** [reply]
@richgcook unfortunately not anymore because of CodeSandbox introducing their 20 sandboxes limit. I had to delete it. Please read this:

- https://github.com/davidjerleke/embla-carousel/discussions/722#discussion-6154094

Also, could you define:

> on a new project neither approach works

What part isn't working?

**Reply by richgcook** [reply]
Okay I'll get one set up and share shortly. Much appreciated.

And yes so the slider all initialises and works fine but the plugins don't run. When I log the embla instance it has a length of 0 for the plugins array.

**Reply by richgcook** [reply]
Here we go, sorry for the delay.

https://stackblitz.com/edit/ab-nuxt3-nhg5fc?file=pages%2Findex.vue

You'll see setup under the components > SliderA.vue for embla. You'll see in the browser window no class names are being added nor is AutoScroll enabled. Weird?

**Reply by meirroth** [reply]
@richgcook Here's a **minimal** setup with Auto Scroll and Class Names plugins **working**: 
https://stackblitz.com/edit/github-gtivz7?file=app.vue

**Reply by richgcook** [reply]
So the issue is using slots? I don't understand. The loop is generated through the component so the slider initialisation should be fine. Hmm.

**Reply by meirroth** [reply, marked as answer]
@richgcook Upgrading `embla-carousel-vue`dependency from `7.1.0` to `8.0.2` fixed the issue 😅
https://stackblitz.com/edit/ab-nuxt3-urc5zc?file=package.json

**Reply by davidjerleke** [reply]
@richgcook probably because of this (from the [`v8.0.0`](https://github.com/davidjerleke/embla-carousel/releases/tag/v8.0.0) release):

<img width="1033" alt="v8" src="https://github.com/davidjerleke/embla-carousel/assets/11529148/d10595a8-1535-437d-bdbe-dac1c982c840">

**Reply by davidjerleke** [reply]
@richgcook please mark @meirroth's comment as answer if it solved your problem. Thanks!

**Reply by richgcook** [reply]
Ah okay so I was using v8 of the plugins but v7 of embla itself so embla wasn't using `ResizeObserver` or `MutationObserver` but the plugins were?

**Reply by davidjerleke** [reply]
@richgcook you need to match the major version of the Embla package with the plugins you use. Example:

- Embla `v7.x.x` only work with plugins with `v7.x.x`.
- Embla `v8.x.x` only work with plugins with `v8x.x`.
- You should **prefer an exact version match** though: Embla `v8.0.3`  with plugins with `v8.0.3`.

This is because v8 has changed how it handled plugins internally.

**Reply by richgcook** [reply]
Thanks so much for the explainer, @davidjerleke – all sorted now :)
